pkg/codegen/dotnet/gen_program_expressions.go

```diff
@@ -350,6 +344,12 @@ func (g *generator) genIntrensic(w io.Writer, from model.Expression, to model.Ty
 	}
 }
 
+func (g *generator) withinAwaitBlock(run func()) {
+	g.insideAwait = true
```

Comment:
I would probably cache the old value, instead of assuming that it is `false`. See
```go
g.withAwaitBlock(func() {
  g.withAwaitBlock(func() {
    contract.Assert(g.insideAwait)
  }
  contract.Assert(g.insideAwait) // panics
}